doc/content/xcp-networkd/host-network-device-ordering.md

```diff
@@ -0,0 +1,181 @@
+---
+title: Host Network Device Ordering
+description: Ordering the host network devices
+---
+
+Purpose
+-------
+
+One of the Toolstack's functions is to maintain a pool of hosts. A pool can be
+constructed by joining a host into an existing pool. One challenge in this
+process is determining which pool-wide network a network device on the joining
+host should connect to.
+
+At first glance, this could be resolved by specifying a mapping between an
+individual network device and a pool-wide network. However, this approach
+would be burdensome for administrators when managing many hosts. It would be
+more efficient if the Toolstack could determine this automatically.
+
+To achieve this, the Toolstack components on two hosts need to independently
+work out consistent identifications for the host network devices and connect
+the network devices with the same identification to the same pool-wide network.
+The identifications on a host can be considered as an order, with each network
+device assigned a unique position in the order as its identification. Network
+devices with the same position will connect to the same network.
+
+
+The assumption
+--------------
+
+Why can the Toolstack components on two hosts independently work out an expected
+order without any communication? This is possible only under the assumption that
+the hosts have identical hardware, firmware, software, and the way
+network devices are plugged into them. For example, an administrator will always
+plug the network devices into the same PCI slot position on multiple hosts if
+they want these network devices to connect to the same network.
+
+The ordering is considered consistent if the positions of such network devices
+(plugged into the same PCI slot position) in the generated orders are the same.
+
+Particularly, when this assumption holds, a consistent initial order can be
+worked out on multiple hosts independently with the help of `biosdevname`. The
+"all_ethN" policy of the `biosdevname` utility can generate a device order based
+on whether the device is embedded or not, PCI cards in ascending slot order, and
+ports in ascending PCI bus/device/function order breadth-first. Since the hosts
+are identical, the orders generated by the `biosdevname` are consistent across
+the hosts.
```

Comment:
I believe there are a few more assumptions that could be stated:

* The order of mac, PCI, eth is total
* NICs have pairwise disjoint MAC addresses
* NICs have pairwise disjoint network (ethX) slots
* Two NICs may have identical PCI slots (but it is rare)
* The sequence of PCI slots in use may have holes
* The sequence of eth slots may have holes

Reply:
I would like to see `biosdevname` explained with an example. What does this look like?

Reply:
> The order of mac, PCI, eth is total

This is true in one enumeration. But this is not in multiple enumerations. E.g. a NIC (mac, PCI, eth) could be used to determine its position in one enumeration, but the PCI or eth might change in the next enumeration. Relying on the total order may result in different positions in different enumerations.

> Two NICs may have identical PCI slots (but it is rare)

This is not rare. But anyway it needs to be supported.

> The sequence of eth slots may have holes

This doesn't have holes.